packages/console/src/contexts/TenantsProvider.tsx

```diff
@@ -1,79 +1,127 @@
 import { defaultManagementApi } from '@logto/schemas';
 import { type TenantInfo, TenantTag } from '@logto/schemas/models';
-import { conditional, noop } from '@silverhand/essentials';
+import { conditionalArray, noop } from '@silverhand/essentials';
 import type { ReactNode } from 'react';
 import { useCallback, useMemo, createContext, useState } from 'react';
 import type { NavigateOptions } from 'react-router-dom';
 
 import { isCloud } from '@/consts/env';
 import { getUserTenantId } from '@/consts/tenants';
 
-type Props = {
-  children: ReactNode;
-};
+type CurrentTenantStatus = 'pending' | 'validating' | 'validated';
 
+/** @see {@link TenantsProvider} for why `useSWR()` is not applicable for this context. */
 type Tenants = {
-  tenants?: TenantInfo[];
-  isSettle: boolean;
-  setTenants: (tenants: TenantInfo[]) => void;
-  setIsSettle: (isSettle: boolean) => void;
+  tenants: readonly TenantInfo[];
+  /** Indicates if the tenants data is ready for the first render. */
+  isInitComplete: boolean;
+  /** Reset tenants to the given value. It will overwrite the current tenants data and set `isInitComplete` to `true`. */
+  resetTenants: (tenants: TenantInfo[]) => void;
+  /** Append a new tenant to the current tenants data. */
+  appendTenant: (tenant: TenantInfo) => void;
+  /** Remove a tenant by ID from the current tenants data. */
+  removeTenant: (tenantId: string) => void;
+  /** Update a tenant by ID if it exists in the current tenants data. */
+  updateTenant: (tenantId: string, data: Partial<TenantInfo>) => void;
+  /**
+   * The current tenant ID that the URL is pointing to. It is navigated programmatically
+   * since there's [no easy way](https://stackoverflow.com/questions/34999976/detect-changes-on-the-url)
+   * to listen to the URL change without polling.
+   */
   currentTenantId: string;
-  setCurrentTenantId: (tenantId: string) => void;
-  navigate: (tenantId: string, options?: NavigateOptions) => void;
+  currentTenant?: TenantInfo;
+  /** Indicates if the Access Token has been validated for use. Will be reset to false when the current tenant changes. */
+  currentTenantStatus: CurrentTenantStatus;
```

Comment:
nit: Seems it's incorrect in the annotation. It would not be reset to false since it's not a boolean flag

Reply:
my bad. was boolean before, will update